Quadriga/src/main/java/edu/asu/spring/quadriga/service/network/impl/NetworkTransformer.java

```diff
@@ -35,45 +28,60 @@
+    private ElementEventTypeDownloadService elementEventTypeDownloadService;
 
     private static final Logger logger = LoggerFactory
             .getLogger(NetworkTransformer.class);
 
+    private static int NUM_THREADS = 4;
+
     /**
      * 
      * {@inheritDoc}
      */
     @Override
     public ITransformedNetwork transformNetwork(
             List<INetworkNodeInfo> networkTopNodesList) {
-        Map<String, Node> nodes = new HashMap<String, Node>();
-        List<Link> links = new ArrayList<Link>();
+        Map<String, Node> nodes = new HashMap<>();
+        List<Link> links = new ArrayList<>();
         ITransformedNetwork transformedNetwork = new TransformedNetwork(nodes, links);
 
-        if (networkTopNodesList != null && networkTopNodesList.size() > 0) {
-            for (INetworkNodeInfo networkNodeInfo : networkTopNodesList) {
-                if (networkNodeInfo.getStatementType().equals(
-                        INetworkManager.RELATIONEVENT)) {
-                    try {
-                        parser.parseStatement(networkNodeInfo.getId(),
-                                nodes, links);
-                    } catch (JAXBException e) {
-                        logger.error("Issue while parsing the JAXB object",
-                                e);
-                    } catch (QStoreStorageException e) {
-                        logger.error("QStore retrieve error", e);
-                    }
-                }
-            }
+        if (networkTopNodesList == null || networkTopNodesList.size() == 0) {
+            // return the trasnformed network
+            return transformedNetwork;
         }
 
+        // networkTopNode list has size > 0
+        // Create executors and tasks
+        ExecutorService executorService = Executors.newFixedThreadPool(NUM_THREADS);
+        // Completion service
+        CompletionService<ElementEventsType> completionService = new
+                ExecutorCompletionService<>(executorService);
+
+        // submit tasks
+        networkTopNodesList.forEach((networkNodeInfo) -> {
+            Callable<ElementEventsType> callable = elementEventTypeDownloadService
+                    .getElementEventTypeDownloadTask(networkNodeInfo.getId());
+            completionService.submit(callable);
+        });
+
+        // check for callabale executions
+        networkTopNodesList.forEach((networkNodeInfo -> {
+            try {
+                Future<ElementEventsType> future = completionService.take();
+                // take() will block if none are executed
+                ElementEventsType elementEventsType = future.get();
+                parser.parseStatement(networkNodeInfo.getId(), elementEventsType, nodes, links);
+            } catch (InterruptedException | ExecutionException e) {
+                logger.error("Issue while retrieving and marshalling object", e);
+            }
+        }));
+        // finally shutdown the service
+        executorService.shutdown();

```

Comment:
Also, let's put all of this code in its own class. The class should have a method that returns a lit of ElementEventsType once all threads are done. Then this method here can call the parser for each element. Does this make sense? This way, the parser does the parsing, your new class does the thread/future handling, and this class just coordinates the transformation. single responsibility principal ;)